### DISCUSSION TITLE
How to avoid deserialization of untrusted data attacks with Hyperion serializer?

### DISCUSSION THREAD

**Discussion by jonnydee** [akkadotnet/akka.net#5329] [Q&A] [answered]
We are currently using Akka 1.4 with Hyperion serialization configured for all user messages sent between Akka systems. We want to make sure Hyperion cannot be tricked into deserialization of untrusted data (see [CWE-502](https://cwe.mitre.org/data/definitions/502.html) for details about that matter).

I know about Hyperion's `akka.actor.serialization-settings.hyperion.disallow-unsafe-type` configuration option which can be used to enable a built-in blacklist for dangerous types. This approach doesn't suffice in our case, however, because we need to trust that the built-in blacklist currently is complete and also will remain complete in the future. Note that being forced to upgrade Akka to some newer release just because new/missing types might have been added to the blacklist is not an option for us. Being able to configure types on the blacklist ourselves would surely be an improvement but the general problem of an outdated blacklist would still remain. It would just be a shift in who is responsible / trusted to have the most recent and complete list of types on the blacklist.

We would like to have an option for a whitelist approach where we list all types that are allowed to be deserialized explicitly. I looked into Hyperion's "Pre-Configured Types" mechanism which involves an [`IKnownTypesProvider`](https://getakka.net/api/Akka.Serialization.IKnownTypesProvider.html). I hoped it would allow us to configure known types explicitly while at the same time ruling out unknown ones. Please correct if I am wrong, but it seems like this is not the case. Instead, known types would be replaced by some id (which is good) but "unknown" types would still be serialized with type information (which is bad). Furthermore, Hyperion's current implementation for mapping ids to known types seems to depend on the order of the types returned by `IKnownTypesProvider.GetKnownTypes()` method. (The id corresponds to the type's position in the returned `IEnumerable<Type>`.) This makes an approach where known message types are collected by some automatism (e.g. a reflection-based approach) difficult, because all involved Actor Systems would have to see exactly the same list (size/items/order) of known types even though some of them would never see all messages in that "overall" list.

We know that using some schema-based serialization technology (e.g. Google's Protocol Buffers) doesn't suffer from such attack vectors but, unfortunately, we cannot use it in our product. Also securing communication between Actor Systems using a VPN is not an option for us.

So the questions are:

- Is there some other way to defend against untrusted data attacks when using Hyperion?
- Do you have any idea on how to implement a whiltelisting machanism for Hyperion in an easy and future-proof way?

**Comment by Aaronontheweb** [comment]
@jonnydee 

First, sorry for the delay in getting back to you.

So right now it looks like our unsafe types list is static when it really should be configurable. @Arkatufus on our team will open an issue for that and we'll work on some solutions.

I like the idea of taking the `IKnownTypesProvider` and passing in a setting that says "only allow these types to be deserialized" - that seems like a robust way of handling this. Alternatively, maybe we should allow users to define a construct similar to a types provider but for the purposes of filtering what can and can't be serialized - i.e. maybe a predicate filter. We do have to balance performance with however we do this, so maybe trying something simple with the `IKnownTypesProvider` would be a good compromise.

**Reply by Aaronontheweb** [reply]
> Furthermore, Hyperion's current implementation for mapping ids to known types seems to depend on the order of the types returned by IKnownTypesProvider.GetKnownTypes() method. (The id corresponds to the type's position in the returned IEnumerable<Type>.) This makes an approach where known message types are collected by some automatism (e.g. a reflection-based approach) difficult, because all involved Actor Systems would have to see exactly the same list (size/items/order) of known types even though some of them would never see all messages in that "overall" list.

FYI, this also seems like a bad Hyperion design if this is true - we should probably move to an explicit K/V system, where the developer must specify the "key identifier" for the type, so known type providers can have their lists extended. We should provide a portability bridge for existing `IKnownTypesProvider` implementations to convert to this without damaging existing wire formats.

**Comment by Arkatufus** [comment, marked as answer]
As of now, the return value of `IKnownTypesProvider.GetKnownType()` is the one being passed to the underlying Hyperion serializer, so it can be considered as immutable after the serializer starts.

Expanding `IKnownTypesProvider` with another method that returns a list of allowed/disallowed types is a viable solution, but it will break compatibility with previous versions of Akka; a safer solution would be to provide another interface to pass in the allowed/disallowed types. It shouldn't matter coding wise, since you can use one class that implements both interface. This implementing class can then be passed in in the HOCON settings just like the known types provider setting.

Adding a toggle in HOCON to let Hyperion to use the list provided by `IKnownTypesProvider.GetKnownType()` as a whitelist can also work, though there might be a case where user would want some flexibility and have other types to be allowed while not listed as a known type, most probably for backward compatibility.

Another option would be to pass a class that contains a delegate callback, but that means we will couple Hyperion and Akka and that just seemed wrong to me,

**Reply by Aaronontheweb** [reply]
> a safer solution would be to provide another interface to pass in the allowed/disallowed types. It shouldn't matter coding wise, since you can use one class that implements both interface. This implementing class can then be passed in in the HOCON settings just like the known types provider setting.

Providing a separate abstraction (i.e. some type of strategy pattern for determining which types are safe or unsafe) for the purposes of filtering makes the most sense - coupling two unrelated things (type hints for performance vs. type restrictions for security) would be a footgun. We can also expand the `HyperionSetup` parameters to make this programmatic, rather than 100% configuration-driven.

We could, in the future, optionally cache the output of this strategy pattern object to speed up include / exclude decisions but let's measure it first.

We should also do the following, upon implementing this:

1. Update https://getakka.net/articles/remoting/security.html to include a section on serialization safety and best practices.
2. Update https://getakka.net/articles/networking/serialization.html#how-to-setup-hyperion-as-default-serializer to include an example of how to use this construct during setup - and honestly, probably move it onto its own page. Lots of Hyperion content now.
3. Reference both of these documents from the Hyperion `README.md`

**Reply by Aaronontheweb** [reply]
We only take a single type filter, although users could certainly compose them if needed. Not our problem.

Some ideas for this abstraction:

**String-based filtering**

```csharp
public interface ITypeFilter{
   public bool Exclude(string typeName); // filter on the name, so we don't have to actually fetch / retrieve the offending type
}
```

**`Type` filtering**

This would allow users to use `Type` metadata to do filtering, i.e. implements X interface.

Worth noting: `LoadTypeByName` already excludes our set of known "evil" types, so this `ITypeFilter` check would be _in addition_ to that.

```csharp
public interface ITypeFilter{
   public bool Exclude(Type typeName); //we've retrieved the type name via LoadTypeByName already
}
```

How I would implement this?

I'd wrap the `LoadTypeByName` call site:

https://github.com/akkadotnet/Hyperion/blob/52ec7921a4d35b6361fe6d98701b8c075fa0e1ad/src/Hyperion/Extensions/TypeEx.cs#L195-L222

With a function that takes the "inner" `LoadTypeByName` output and passes it through the filter. If the `Exclude` function return `true`, we throw a derivative of the `EvilDeserializationException` that makes it clear that this type was determined to be "evil" by user-code, not by Hyperion's own built-in screening system. The end-result is the same, but it should stand out as distinct inside the user's error logs.

To implement a whitelist-style of filtering with this system, you'd just want to implement the `Exclude` function with a set of booleans for whatever you want to include and then just `!` it at the end. Would that work for your use case @jonnydee ?

**Reply by Aaronontheweb** [reply]
I think the string-based filtering is the way to go for best security - that way we can entirely avoid the issue of having to parse the `Type` and load the assembly it's in to perform reflection. This is going to have some performance impact (i.e. I imagine a lot of users are going to want to use `string.Contains` et al to validate that these types belong to a specific assembly.).

Another way of doing this, which might be more secure, is to use a builder pattern that requires the user to specify what _is_ allowed:

```csharp
var myFilter = TypeFilterBuilder.Create()
   .Include<TMessage1>()
   .Include<TMessage2>()
   .Include<TMessage3>();
   .Build();
```

This would, in effect, grab the `string` representations of these types used by Hyperion and feed them into the `ITypeFilter`. Any messages with an exact match would be included. We're not going to check subtype / marker interface, gobbledygook - if you want to secure your Hyperion serialization to the maximum degree possible, build this class to include all of the verbatim type names you will allow. Full stop.

**Reply by Aaronontheweb** [reply]
I think we should combine the `ITypeFilter` with a string implementation along with a statically typed builder interface and that's what we should recommend in our literature. If users want to do weird stuff with `string.Contains` inside an `ITypeFilter` we'll let them do that but tell them in the documentation "the right way to do this is with the builder."

We'll also encourage them to use the `HyperionSerializerSetup` class instead of HOCON for configuring this so this configuration is all compile-time checked prior to starting the `ActorSystem`.

**Reply by Aaronontheweb** [reply]
Before we do any implementation work on this, let's leave this open for end-user and contributor comment for 3-4 days. Want some feedback in case we're barking up the wrong tree. cc @jonnydee

**Reply by to11mtm** [reply]
@Aaronontheweb  my comments:


> We're not going to check subtype / marker interface, gobbledygook - if you want to secure your Hyperion serialization to the maximum degree possible, build this class to include all of the verbatim type names you will allow

Frankly if you wanted such functionality it would be fairly trivial to implement as an extension method. However it is a footgun if people don't understand the implications of using such a thing (i.e. you'd only want it on interfaces within your own private codebase).

One plus side of the builder pattern alongside using strings, is frankly it makes implementation somewhat clean: If we are in 'explicit' mode and specifying all types up front, that means we could potentially in that mode, branch on loading type by name; explicit mode could get to use a normal dictionary string-type cache, instead of the existing concurrent dictionary (since all types are preadded, don't have to worry about the 'add' case being concurrent.) So you'd basically pay a heavier penalty for creating the serializer, but overall deserialization could actually be -faster-!

**Reply by jonnydee** [reply]
@Aaronontheweb Thank you very much for your response and suggested solutions. Personally, I like the string-based filtering approach together with the 'TypeFilterBuilder'. However, I'm not sure if my initial 'whiltelist' idea was a good one, because it might be tedious and error prone to list all known message types. Your suggested 'ITypeFilter' interface with 'Exclude' method seems to go for the blacklisting approach which is indeed better, I think. Correspondingly, the 'TypeFilterBuilder' should provide an '.Exclude<T>' method which would be used to build up the blacklist in a type-safe manner.

Of course, being able to either backlist OR whitelist types would allow for greatest flexibility.

**Reply by Aaronontheweb** [reply]
@jonnydee with that exclusion system you could still implement "whitelist"-style filtering by simply taking everything in your filter you want to include and then just `!`-ing it inside the boolean logic whereever that goes. "Anything that is not one of these types gets excluded"

This would allow you to specify other conditions too, such as "anything that is not on this list" AND "anything on this list that doesn't belong to an Assembly with name X" or whatever. 

It will be tedious - you'd have to pre-scan the types you'd want to include using some assembly scanning on the builder, but that way the reflection is controlled by _you_ when you're setting up the filter, not by the filter itself when it's evaluating messages.

**Reply by jonnydee** [reply]
@Aaronontheweb You're right, of course. Thanks again for your help. The chosen approach is brilliant! :-)

**Reply by Aaronontheweb** [reply]
No problem! We're getting ready to merge it into Hyperion and release it soon: https://github.com/akkadotnet/Hyperion/pull/281

With any luck that should be available by the end of the week and we'll also adapt Akka.Serialization.Hyperion to use it.

**Comment by Aaronontheweb** [comment]
@Arkatufus has posted a full spec for this here: https://github.com/akkadotnet/Hyperion/issues/275#issuecomment-994900073

**Reply by Aaronontheweb** [reply]
This is now merged into Akka.NET v1.4.32 https://github.com/akkadotnet/akka.net/pull/5510

This change will be available in our nightlies beginning tomorrow: https://getakka.net/community/getting-access-to-nightly-builds.html

And we'll release 1.4.32 soon.

**Reply by Aaronontheweb** [reply]
cc @jonnydee

**Reply by Aaronontheweb** [reply]
@jonnydee this was released as part of Akka.NET v1.4.32 yesterday.

**Reply by jonnydee** [reply]
@Aaronontheweb thanks for this great support and for letting me know 😀

**Reply by Aaronontheweb** [reply]
No problem. Let us know if you run into any issues with it.
